### PR TITLE
Unit tests for machineset controller

### PR DESCRIPTION
**Pull request by hardikdr**
**What this PR does / why we need it**:  This PR adds few tests for machine-set.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```improvement operator
Enhanced unit-test coverage for machineset controller
```

### CODE REVIEW

File: pkg/controller/machineset_test.go [hunk 1]
```diff
@@ -0,0 +1,1022 @@
...
+				Status: machinev1.MachineStatus{
+					CurrentStatus: machinev1.CurrentStatus{
+						Phase: MachineFailed,
+					},
+				},
+			}
+		})
+
+		// Testcase: It should delete the inactive machines.
+		It("should not return error", func() {
```
Comment (amshuman-kr) [inline pkg/controller/machineset_test.go]:
```suggestion
		It("It should delete the inactive machines.", func() {
```

File: pkg/controller/machineset_test.go [hunk 2]
```diff
@@ -0,0 +1,1022 @@
...
+							"test-label": "test-label",
+						},
+					},
+				},
+			}
+
+		})
+
+		// Testcase: It should add finalizer on MachineSet.
+		It("should not return error", func() {
```
Comment (amshuman-kr) [inline pkg/controller/machineset_test.go]:
```suggestion
		It("should add finalizer on MachineSet.", func() {
```

File: pkg/controller/machineset_test.go [hunk 3]
```diff
@@ -0,0 +1,1022 @@
...
+			c, trackers := createController(stop, testNamespace, objects, nil, nil)
+			defer trackers.Stop()
+			waitForCacheSync(stop, c)
+
+			c.addMachineSetFinalizers(testMachineSet)
+
+			waitForCacheSync(stop, c)
+			testMachineSet, _ := c.controlMachineClient.MachineSets(testNamespace).Get(testMachineSet.Name, metav1.GetOptions{})
+
+			Expect(testMachineSet.Finalizers[0]).To(Equal(DeleteFinalizerName))
```
Comment (amshuman-kr) [inline pkg/controller/machineset_test.go]:
```suggestion
			Expect(testMachineSet.Finalizers).To(HaveLen(1))
			Expect(testMachineSet.Finalizers).To(ContainElement(DeleteFinalizerName))
```

File: pkg/controller/machineset_test.go [hunk 4]
```diff
@@ -0,0 +1,1022 @@
...
+							"test-label": "test-label",
+						},
+					},
+				},
+			}
+
+		})
+
+		// Testcase: It should delete the finalizer from MachineSet.
+		It("should not return error", func() {
```
Comment (amshuman-kr) [inline pkg/controller/machineset_test.go]:
```suggestion
		It("should delete the finalizer from MachineSet.", func() {
```

File: pkg/controller/machineset_test.go [hunk 5]
```diff
@@ -0,0 +1,1022 @@
...
+			objects := []runtime.Object{}
+			objects = append(objects, testMachineSet)
+			c, trackers := createController(stop, testNamespace, objects, nil, nil)
+			defer trackers.Stop()
+			waitForCacheSync(stop, c)
+
+			c.deleteMachineSetFinalizers(testMachineSet)
+
+			waitForCacheSync(stop, c)
+			testMachineSet, _ := c.controlMachineClient.MachineSets(testNamespace).Get(testMachineSet.Name, metav1.GetOptions{})
```
Comment (amshuman-kr) [inline pkg/controller/machineset_test.go]:
```suggestion
			testMachineSet, _ = c.controlMachineClient.MachineSets(testNamespace).Get(testMachineSet.Name, metav1.GetOptions{})
```

File: pkg/controller/machineset_test.go [hunk 6]
```diff
@@ -0,0 +1,1022 @@
...
+			stop := make(chan struct{})
+			defer close(stop)
+
+			objects := []runtime.Object{}
+			objects = append(objects, testMachineSet)
+			c, trackers := createController(stop, testNamespace, objects, nil, nil)
+			defer trackers.Stop()
+			waitForCacheSync(stop, c)
+
+			c.deleteMachineSetFinalizers(testMachineSet)
```
Comment (amshuman-kr) [inline pkg/controller/machineset_test.go]:
```suggestion
			testMachineSet, _ := c.controlMachineClient.MachineSets(testNamespace).Get(testMachineSet.Name, metav1.GetOptions{})
			Expect(testMachineSet.Finalizers).Should(Not(BeEmpty()))
			
			c.deleteMachineSetFinalizers(testMachineSet)
```

File: pkg/controller/machineset_test.go [hunk 7]
```diff
@@ -0,0 +1,1022 @@
...
+					},
+				},
+			}
+
+			finalizers = []string{"finalizer1", "finalizer2"}
+
+		})
+
+		// Testcase: It should update the finalizer on MachineSet.
+		It("should not return error", func() {
```
Comment (amshuman-kr) [inline pkg/controller/machineset_test.go]:
```suggestion
		It("should update the finalizer on MachineSet.", func() {
```

File: pkg/controller/controller_utils.go [hunk 8]
```diff
@@ -621,6 +623,112 @@ func (r RealMachineControl) DeleteMachine(namespace string, machineID string, ob
...
+func (r FakeMachineControl) CreateMachinesWithControllerRef(namespace string, template *v1alpha1.MachineTemplateSpec, controllerObject runtime.Object, controllerRef *metav1.OwnerReference) error {
+	if err := validateControllerRef(controllerRef); err != nil {
+		return err
+	}
+	return r.createMachines(namespace, template, controllerObject, controllerRef)
+}
+
+// PatchMachine applies a patch on machine
+func (r FakeMachineControl) PatchMachine(namespace string, name string, data []byte) error {
+	_, err := r.controlMachineClient.Machines(namespace).Patch(name, types.MergePatchType, data)
```
[the remark this answers is not in the document]
Comment (hardikdr) [reply pkg/controller/controller_utils.go]: I need to enable recording in general, will do soon.

File: pkg/controller/machineset_test.go [hunk 9]
```diff
@@ -0,0 +1,1022 @@
...
+
+			testMachine := &machinev1.Machine{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "Machine-test",
+					Namespace: testNamespace,
+					Labels:    nil,
+				},
+			}
+			MachineSet, err := c.getMachineMachineSets(testMachine)
+			Expect(err).Should(Not(BeNil()))
```
[the remark this answers is not in the document]
Comment (hardikdr) [reply pkg/controller/machineset_test.go]: done.

File: pkg/controller/machineset_test.go [hunk 10]
```diff
@@ -0,0 +1,1022 @@
...
+			c, trackers := createController(stop, testNamespace, objects, nil, nil)
+			defer trackers.Stop()
+			waitForCacheSync(stop, c)
+
+			MachineSet := c.resolveMachineSetControllerRef(testNamespace, testControllerRef)
+			Expect(MachineSet).Should(Not(BeNil()))
+		})
+
+		// Testcase: It should return MachineSet if name and UID matches.
+		It("should not return MachineSet", func() {
```
[the remark this answers is not in the document]
Comment (hardikdr) [reply pkg/controller/machineset_test.go]: Yes, I will adapt to it while adding new tests.

**Comment by prashanth26** [comment]
Checks seem to be failing due to the package issue that is fixed on master, and tests with AWS failed due to some unexpected timeout issues. Seems to be safe to merge.

**Pull request merged**